### ISSUE TITLE
Out of the box adaptive learning rates: Feature idea

### ISSUE THREAD

**Issue by benjohnston24** [Lasagne/Lasagne#722]
Hello Lasagnians...

I have a feature request / idea.  Hopefully it isnt too trivial / superficial.  I thought it would be handy to include some commonly used adaptive learning rate options out of the box; perhaps into updates.py.  The idea being that after an update step is executed, the corresponding update function e.g. sgd calls learningrate() prior to exit to change the learning rate; maybe wrapping the learning rate in a decorator.  You could use the existing update api but instead of providing a single value, provide an instance of a class e.g.

In updates.py we could have something similar to
```
class learningrate(object):
    def __init__(self, start):
        self.learning_rate = start
      ...
    def __call__(self):
        self.learning_rate = some change to the rate
        return self.learning_rate

def learning_rate_decorator(update):
    def update_wrapper(loss, params, learning_rate, ...):
        update(loss, params, learning_rate, ....)
        
         # Update the learning rate
         if isinstance(learning_rate, learningrate):
             learning_rate()

.....
@learning_rate_decorator
def sgd(loss, params.....):
         
```
And when we call the code something like

```
from updates import lineardecay, sgd # lineardecay inherits from the learningrate class

...
learn_rate = lineardecay(rate=0.2)
updates_sgd = sgd(loss, params, learning_rate=learn_rate)

# Now learn_rate should update automatically, being reduced after sgd executes
```

A base LearningRate class could be extended to include different types of adaptive learning rates e.g. exponential, linear decay, etc.  TensorFlow does implement something similar but only with the option for exponential decay [link here](https://www.tensorflow.org/versions/r0.7/api_docs/python/train.html#exponential_decay).  I would be very happy to write / help on the implementation, though I am a newbie when it comes to contributing to open source projects so any guidance would be appreciated.

I am interested to hear your thoughts

Thanks in advance,
Ben

**Comment by f0k** [comment]
Thanks for your suggestion! Unfortunately, it has a small problem: The learning rate is usually adapted after a fixed number of updates (e.g., an epoch), not after every single update, so this has to be maintained in the training loop and not in the update dictionary.

If you really want to decay the learning rate after each update, it's as simple as:
```python
eta = theano.shared(lasagne.utils.floatX(initial_eta))
updates = lasagne.updates.sgd(loss, params, learning_rate=eta)
updates[eta] = eta * 0.999
```

Or:
```python
eta = theano.shared(lasagne.utils.floatX(initial_eta))
eta.default_update = eta * 0.999
updates = lasagne.updates.sgd(loss, params, learning_rate=eta)
```

This would not really get any simpler by providing a wrapper (and as I said, the use would be limited, since you'd typically not change the rate after every update step). Automating this only makes sense once (and if) we add support for a training loop. Until then, you should just manually decay the learning rate in your training loop:
```python
eta = theano.shared(lasagne.utils.floatX(initial_eta))
updates = lasagne.updates.sgd(loss, params, learning_rate=eta)
...
for ... in ...:
    for ... in ...:
        ... = train_fn(...)
    eta.set_value(eta.get_value() * lasagne.utils.floatX(0.98))
```
Hope this helps!

PS: You should go through the [Theano tutorial](http://deeplearning.net/software/theano/tutorial/), or the [extended Lasagne tutorial](http://f0k.de/lasagne-embl/) -- you fell for a common misunderstanding about what is executed when in symbolic programming. In particular, your `self.learning_rate = some change to the rate` would only be executed once, not after every update.

**Comment by benjohnston24** [comment]
Hi f0k,

Thanks for the detailed response!  I have used the theano shared variables for updating the learning rate, but totally forgot about it being bound to the training loop.  Thanks for the tips, I'm a huge fan of the project and hope to be able to help in some way soon.

Thanks,
Ben
